docs/proposals/source-verificiation-policies.md

```diff
@@ -0,0 +1,313 @@
+
+    Since every commit between the last synced revision and the target revision requires to be signed, an adversary modifying the `Application` CR in a non-intended way could only ever move _backwards_ in time (i.e. take values from a commit that has synced before), there would be no way to slip in an unsigned commit somewhere.
+
+    **Risk:** Using the progressive verification level could allow an adversary to sync 
+    
+2. The `progressive` level has a "bootstrap" feature to allow creating an `Application` in `progressive` level. Obviously, when an `Application` is created, it does not have any history information unless it is synced initially. The bootstrap feature defines a time window, the `bootstrapPeriod`, in which an `Application` without history can temporarily sync using the `head` level (i.e. only with the HEAD of `targetRevision` being signed). The bootstrap feature assumes that the `.metadata.creationTimestamp` of the `Application` resource is immutable, or rather be reset by the Kubernetes API server upon modification. During the timeframe in the `bootstrapPeriod`, an adversary could delete the `Application`'s history and sync to any signed commit in the repository.
+
+3. The `progressive` level doesn't allow roll-backs, as it verifies the commits between the previous synced state and the target revision. If the application is on `F`, and wants to roll back to `C`, it cannot find any revision history for this path (because `C` existed prior to `F`). This must be properly documented. For `head` and `full`, this is not a concern as long as the level's requirements are met in the repository.
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+
+Upgrade will be seamless. When Argo CD detects the prior configuration for Git commit verification (i.e. `.spec.signatureKeys` is populated in the `AppProject`), the new implementation will behave like the current one. 
+
+Internally, this will create a single verification policy similar to the following and ignore all other configured policies: 
+
+```
+repositoryPattern: '*'
+repositoryType: git
+verificationLevel: head
+verificationMethod: gpg
+trustedSigners:
+- # taken from .spec.signatureKeys
+```
+
+If a user wants to migrate from the current implementation to the new source verification policies, they will first have to remove `.spec.signatureKeys` and can then go ahead and define desired policies in `.spec.sourceVerificationPolicies`.
```

Comment:
What's the behavior if both fields are set at the same time?

Reply:
In the current PoC code, as long as `.spec.signatureKeys` is set, it will take precedence.